### PR TITLE
Various tweaks related to compilation. Fixes #3390

### LINKED ISSUE #3390: Errors trying to build, run Java runtime tests on fresh M1 mac with java 16
I get with this error via command line:

```
$ mvn -DskipTests clean install
...
Java version: 16.0.2, vendor: Oracle Corporation, runtime: /Library/Java/JavaVirtualMachines/jdk-16.0.2.jdk/Contents/Home
...
[INFO] ANTLR 4 Runtime Tests (2nd generation) ............. FAILURE [ 12.637 s]
...
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project antlr4-runtime-testsuite: Fatal error compiling: java.lang.IllegalAccessError: class org.antlr.v4.test.runtime.CommentHasStringValueProcessor (in unnamed module @0x2c7df4fa) cannot access class com.sun.tools.javac.model.JavacElements (in module jdk.compiler) because module jdk.compiler does not export com.sun.tools.javac.model to unnamed module @0x2c7df4fa -> [Help 1]
```

Adding a series of these to `maven-compiler-plugin` from this [doc](https://errorprone.info/docs/installation#jdk-16) doesn't help:

```
<configuration>
      ...
      <fork>true</fork>
      <compilerArgs>
                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED</arg>
```

### PR DESCRIPTION
**Pull request by parrt**
Please take a look @KvanTTT and @ericvergnaud Fixes #3390 ?

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -155,6 +155,19 @@
...
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED</arg>
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED</arg>
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED</arg>
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED</arg>
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED</arg>
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED</arg>
+                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
+                            <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED</arg>
+                            <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED</arg>
```
Comment (KvanTTT) [inline pom.xml:169]: It seems like some lines are excess. Could you please try these ones?

```suggestion
                            <arg>--add-exports</arg><arg>jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
                            <arg>--add-exports</arg><arg>jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
                            <arg>--add-exports</arg><arg>jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED</arg>
                            <arg>--add-exports</arg><arg>jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED</arg>
```

Also, why do you need `<fork>true</fork>`?

Comment (parrt) [reply pom.xml:169]: I am just going by what this page says is required for Java 16:

https://errorprone.info/docs/installation#jdk-16

Comment (parrt) [reply pom.xml:169]: I am trying that's upset. Interesting. Now I'm getting:

```
[ERROR] error: option --add-exports not allowed with target 7
```

Comment (parrt) [reply pom.xml:169]: Confirmed. It works with that subset. Updating the PR.

**Comment by parrt** [comment]
Still can't get this to build:

```
[INFO] Reactor Summary for ANTLR 4 4.9.4-SNAPSHOT:
[INFO] 
[INFO] ANTLR 4 ............................................ SUCCESS [  0.714 s]
[INFO] ANTLR 4 Runtime .................................... SUCCESS [  4.498 s]
[INFO] ANTLR 4 Tool ....................................... SUCCESS [ 17.222 s]
[INFO] ANTLR 4 Maven plugin ............................... SUCCESS [  8.509 s]
[INFO] ANTLR 4 Runtime Test Annotations ................... SUCCESS [  0.449 s]
[INFO] ANTLR 4 Runtime Test Processors .................... SUCCESS [  0.048 s]
[INFO] ANTLR 4 Runtime Tests (2nd generation) ............. FAILURE [  5.169 s]
[INFO] ANTLR 4 Tool Tests ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  37.675 s
[INFO] Finished at: 2021-12-05T12:51:54-08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project antlr4-runtime-testsuite: Compilation failure -> [Help 1]
```

**Comment by parrt** [comment]
Dang. It looks like the new compiler argument only work with java 16 not 8, but maybe we can require 16 for building the tool, but 8 for using the generated antlr4 jar?

**Comment by KvanTTT** [comment]
I would vote for the latest Java LTS (17). Actually, Java 16 is already not supported.

**Comment by ericvergnaud** [comment]
Sorry but I don’t think we should force antlr users to switch to Java 16, especially when there is no added value in doing so
I’m all for dropping Java 7 and the tool could migrate to a recent version, but certainly not the runtime 

Envoyé de mon iPhone

**Comment by parrt** [comment]
Yes, the runtime should stay java 8 but the build of tool itself could be up to date like 16.

**Comment by KvanTTT** [comment]
16 is not actual, see https://en.wikipedia.org/wiki/Java_version_history

**Comment by parrt** [comment]
I just meant what I could download easily.

**Pull request closed without merging**